propkatraj/propkatraj.py

```diff
@@ -2,22 +2,239 @@
+            wmsg = ("Non protein atoms passed to propka 3.1.\n MDAnalysis' "
+                    "PDBWriter does not currently write non-standard residues "
+                    "correctly as HETATM records and this may lead to "
+                    "incorrect pKa predictions.\n"
+                    "See https://github.com/Becksteinlab/propkatraj/issues/24 "
+                    " for more details")
+            warnings.warn(wmsg)
+
+        # TODO: probably needs to be a temporary directory instead
+        self.tmpfile = os.path.join(os.path.curdir, 'current.pdb')
+        self.skip_failure = skip_failure
+        super(PropkaTraj, self).__init__(self.ag.universe.trajectory, **kwargs)
+
+    def _prepare(self):
+        self._pkas = []
+        self.num_failed_frames = 0
+        self.failed_frames_log = []
+        self.failed_times = []
+        self._columns = None
+
+    def _single_frame(self):
+        pstream = mda.lib.util.NamedStream(StringIO(), self.tmpfile)
+        self.ag.write(pstream)
+        # reset stream for reading
+        pstream.reset()
+
+        try:
+            # TODO: it would be nice to allow for other options, maybe for 3.2?
+            mol = pk.single(pstream, optargs=['--quiet'])
+        except (IndexError, AttributeError) as err:
+            errmsg = "failure on frame: {0}".format(self._ts.frame)
+            if not self.skip_failure:
+                raise_from(RuntimeError(errmsg), err)
+            else:
+                warnings.warn(errmsg)
+                self.num_failed_frames += 1
+                self.failed_frames_log.append(self._ts.frame)
+                self.failed_times.append(self._ts.time)
+        else:
+            confname = mol.conformation_names[0]
+            conformation = mol.conformations[confname]
+            groups = conformation.get_titratable_groups()
+
+            # extract pka estimates from each residue
+            self._pkas.append([g.pka_value for g in groups])
+            if self._columns is None:
+                self._columns = [g.atom.resNumb for g in groups]
+        finally:
+            # deallocate stream
+            pstream.close(force=True)
+
+    def _conclude(self):
+        # to allow for popping times later, convert self.times to list
+        times = self.times.tolist()
+
+        # Ouput failed frames
+        if self.num_failed_frames > 0:
+            perc_failure = (self.num_failed_frames / self.n_frames) * 100
+            # if frames have failed we need to ammend times accordingly
+            for i in self.failed_times:
+                times.remove(i)
+
+            wmsg = ("number of failed frames = {0}\n"
+                    "percentage failure = {1}\n"
+                    "failed frames: {2}".format(self.num_failed_frames,
+                                                perc_failure,
+                                                self.failed_frames_log))
+            logging.warning(wmsg)
```

Comment:
I think you only need to convert times to a list _if_ you modify it.

Furthermore, for consistency I would change times to an array again after modification. I think if anyone wants to poke around in the Analysis object they'd expect to find an array in times.